# Data Understanding

### Dataset Columns Description



| Column Name | Description | 
|-------------|-------------|
| `index` | Sequential row identifier for each record in the dataset. | 
| `transaction_id` | Unique identifier assigned to each sales transaction. |  
| `transaction_date` | Date when the transaction occurred. |  
| `transaction_time` | Exact time when the transaction was processed. |  
| `store_id` | Unique identifier for the coffee shop branch/store. |  
| `store_location` | Physical location or branch name where the sale occurred. |  
| `product_id` | Unique identifier for the sold product. | 
| `transaction_qty` | Quantity of product units purchased in the transaction. |  
| `unit_price` | Price per single unit of the product. |  
| `Total_Bill` | Total amount paid for the transaction (`transaction_qty × unit_price`). |  
| `product_category` | Broad category of the sold product. |  
| `product_type` | Specific type or subcategory of the product. |  
| `product_detail` | Detailed product name or flavor. |  
| `Size` | Product serving size. | String | Large |
| `Month Name` | Full month name extracted from transaction date for easier analysis. |  
| `Day Name` | Day of the week extracted from transaction date. |  
| `Hour` | Hour of transaction extracted from transaction time. |  
| `Month` | Numeric month extracted from transaction date. | 
| `Day of Week` | Numeric representation of weekday (`0=Monday`, `6=Sunday`). | 

### Notes:
- **Primary business metrics:** `transaction_qty`, `unit_price`, `Total_Bill`
- **Time-based analysis columns:** `transaction_date`, `transaction_time`, `Month Name`, `Day Name`, `Hour`
- **Product analysis columns:** `product_category`, `product_type`, `product_detail`, `Size`
- **Store performance columns:** `store_id`, `store_location`

# Data Exploration

In [195]:
import numpy as np
import pandas as pd
import plotly.express as px

In [196]:
pd.set_option('display.max_columns', None)

In [197]:
df = pd.read_csv('Coffee_shop_sales.csv')
df

,transaction_id,transaction_date,transaction_time,store_id,store_location,product_id,transaction_qty,unit_price,Total_Bill,product_category,product_type,product_detail,Size,Month Name,Day Name,Hour,Month,Day of Week
0,114301,01-06-2023,11:33:29,3,Astoria,45,1,3.00,3.0,Tea,Brewed herbal tea,Peppermint,Large,June,Thursday,11,6,3
1,115405,02-06-2023,11:18:24,3,Astoria,45,1,3.00,3.0,Tea,Brewed herbal tea,Peppermint,Large,June,Friday,11,6,4
2,115478,02-06-2023,12:02:45,3,Astoria,45,1,3.00,3.0,Tea,Brewed herbal tea,Peppermint,Large,June,Friday,12,6,4
3,116288,02-06-2023,19:39:47,3,Astoria,45,1,3.00,3.0,Tea,Brewed herbal tea,Peppermint,Large,June,Friday,19,6,4
4,116714,03-06-2023,12:24:57,3,Astoria,45,1,3.00,3.0,Tea,Brewed herbal tea,Peppermint,Large,June,Saturday,12,6,5
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
149111,129465,14-06-2023,08:34:10,5,Lower Manhattan,41,4,4.25,17.0,Coffee,Barista Espresso,Cappuccino,Large,June,Wednesday,8,6,2
149112,133523,17-06-2023,09:55:47,8,Hell's Kitchen,8,8,45.00,360.0,Coffee beans,Premium Beans,Civet Cat,Not Defined,June,Saturday,9,6,5
149113,133674,17-06-2023,10:41:11,8,Hell's Kitchen,8,8,45.00,360.0,Coffee beans,Premium Beans,Civet Cat,Not Defined,June,Saturday,10,6,5
149114,133744,17-06-2023,11:18:31,8,Hell's Kitchen,8,8,45.00,360.0,Coffee beans,Premium Beans,Civet Cat,Not Defined,June,Saturday,11,6,5


### Check Data Types

In [198]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149116 entries, 0 to 149115
Data columns (total 18 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   transaction_id    149116 non-null  int64  
 1   transaction_date  149116 non-null  object 
 2   transaction_time  149116 non-null  object 
 3   store_id          149116 non-null  int64  
 4   store_location    149116 non-null  object 
 5   product_id        149116 non-null  int64  
 6   transaction_qty   149116 non-null  int64  
 7   unit_price        149116 non-null  float64
 8   Total_Bill        149116 non-null  float64
 9   product_category  149116 non-null  object 
 10  product_type      149116 non-null  object 
 11  product_detail    149116 non-null  object 
 12  Size              149116 non-null  object 
 13  Month Name        149116 non-null  object 
 14  Day Name          149116 non-null  object 
 15  Hour              149116 non-null  int64  
 16  Month             14

### Check Summary Statistics for Numerical columns

In [199]:
df.describe().round(2)

,transaction_id,store_id,product_id,transaction_qty,unit_price,Total_Bill,Hour,Month,Day of Week
count,149116.00,149116.00,149116.00,149116.00,149116.00,149116.00,149116.00,149116.00,149116.00
mean,74737.37,5.34,47.92,1.44,3.38,4.69,11.74,3.99,2.98
std,43153.60,2.07,17.93,0.54,2.66,4.23,3.76,1.67,2.00
min,1.00,3.00,1.00,1.00,0.80,0.80,6.00,1.00,0.00
25%,37335.75,3.00,33.00,1.00,2.50,3.00,9.00,3.00,1.00
50%,74727.50,5.00,47.00,1.00,3.00,3.75,11.00,4.00,3.00
75%,112094.25,8.00,60.00,2.00,3.75,6.00,15.00,5.00,5.00
max,149456.00,8.00,87.00,8.00,45.00,360.00,20.00,6.00,6.00


### Check Summary Statistics for Categorical columns

In [200]:
df.describe(include= ['object'])

,transaction_date,transaction_time,store_location,product_category,product_type,product_detail,Size,Month Name,Day Name
count,149116,149116,149116,149116,149116,149116,149116,149116,149116
unique,181,25762,3,9,29,45,4,6,7
top,19-06-2023,09:31:15,Hell's Kitchen,Coffee,Brewed Chai tea,Our Old Time Diner Blend,Regular,June,Friday
freq,1343,41,50735,58416,17183,8660,45789,35352,21701


### Check Duplicates

In [201]:
df.duplicated().sum()

np.int64(0)

### Check Missing Values

In [202]:
df.isna().sum()

transaction_id      0
transaction_date    0
transaction_time    0
store_id            0
store_location      0
product_id          0
transaction_qty     0
unit_price          0
Total_Bill          0
product_category    0
product_type        0
product_detail      0
Size                0
Month Name          0
Day Name            0
Hour                0
Month               0
Day of Week         0
dtype: int64

# Data Cleaning

In [203]:
# Drop unnecessary columns
df.drop(columns= ['store_id', 'product_id'], inplace= True)

In [204]:
df.drop(columns='product_detail', inplace=True)

In [205]:
df.columns

Index(['transaction_id', 'transaction_date', 'transaction_time',
       'store_location', 'transaction_qty', 'unit_price', 'Total_Bill',
       'product_category', 'product_type', 'Size', 'Month Name', 'Day Name',
       'Hour', 'Month', 'Day of Week'],
      dtype='object')

In [206]:
df.duplicated().sum()

np.int64(0)

In [207]:
df.head(1)

,transaction_id,transaction_date,transaction_time,store_location,transaction_qty,unit_price,Total_Bill,product_category,product_type,Size,Month Name,Day Name,Hour,Month,Day of Week
0,114301,01-06-2023,11:33:29,Astoria,1,3.0,3.0,Tea,Brewed herbal tea,Large,June,Thursday,11,6,3


In [208]:
# Convert to datetime
df['transaction_date'] = df['transaction_date'] + ' ' + df['transaction_time']
df['transaction_date'] = pd.to_datetime(df['transaction_date'], format = '%d-%m-%Y %H:%M:%S')

In [209]:
df.drop(columns= ['transaction_time'], inplace= True)
df.head(1)

,transaction_id,transaction_date,store_location,transaction_qty,unit_price,Total_Bill,product_category,product_type,Size,Month Name,Day Name,Hour,Month,Day of Week
0,114301,2023-06-01 11:33:29,Astoria,1,3.0,3.0,Tea,Brewed herbal tea,Large,June,Thursday,11,6,3


In [210]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149116 entries, 0 to 149115
Data columns (total 14 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   transaction_id    149116 non-null  int64         
 1   transaction_date  149116 non-null  datetime64[ns]
 2   store_location    149116 non-null  object        
 3   transaction_qty   149116 non-null  int64         
 4   unit_price        149116 non-null  float64       
 5   Total_Bill        149116 non-null  float64       
 6   product_category  149116 non-null  object        
 7   product_type      149116 non-null  object        
 8   Size              149116 non-null  object        
 9   Month Name        149116 non-null  object        
 10  Day Name          149116 non-null  object        
 11  Hour              149116 non-null  int64         
 12  Month             149116 non-null  int64         
 13  Day of Week       149116 non-null  int64         
dtypes: d

### In depth check for numerical columns

In [211]:
num_cols = df.select_dtypes(include= 'number').columns.drop('transaction_id')
num_cols

Index(['transaction_qty', 'unit_price', 'Total_Bill', 'Hour', 'Month',
       'Day of Week'],
      dtype='object')

In [212]:
for col in num_cols:
    px.box(data_frame= df, x= col, title= col).show()

#### There is some columns that have outliers. 

### In depth check for categorical columns

In [213]:
cat_cols = df.select_dtypes(include= 'object').columns
cat_cols

Index(['store_location', 'product_category', 'product_type', 'Size',
       'Month Name', 'Day Name'],
      dtype='object')

In [214]:
for col in cat_cols:

    print(col)
    print(df[col].nunique())
    print(df[col].unique())
    print('-' * 100)

store_location
3
['Astoria' 'Lower Manhattan' "Hell's Kitchen"]
----------------------------------------------------------------------------------------------------
product_category
9
['Tea' 'Coffee' 'Bakery' 'Drinking Chocolate' 'Flavours' 'Loose Tea'
 'Packaged Chocolate' 'Branded' 'Coffee beans']
----------------------------------------------------------------------------------------------------
product_type
29
['Brewed herbal tea' 'Brewed Black tea' 'Brewed Green tea'
 'Brewed Chai tea' 'Drip coffee' 'Organic brewed coffee'
 'Barista Espresso' 'Gourmet brewed coffee' 'Scone' 'Pastry'
 'Premium brewed coffee' 'Hot chocolate' 'Biscotti' 'Sugar free syrup'
 'Regular syrup' 'Herbal tea' 'Black tea' 'Chai tea' 'Green tea'
 'Drinking Chocolate' 'Organic Chocolate' 'Clothing' 'Housewares'
 'Espresso Beans' 'Green beans' 'Organic Beans' 'Premium Beans'
 'Gourmet Beans' 'House blend Beans']
----------------------------------------------------------------------------------------------------


# EDA & Visualisation

In [215]:
df.head(1)

,transaction_id,transaction_date,store_location,transaction_qty,unit_price,Total_Bill,product_category,product_type,Size,Month Name,Day Name,Hour,Month,Day of Week
0,114301,2023-06-01 11:33:29,Astoria,1,3.0,3.0,Tea,Brewed herbal tea,Large,June,Thursday,11,6,3


#### What is the total revenue per day of the week?

In [216]:
df_revperday = df.groupby(['Day of Week','Day Name'])['Total_Bill'].sum().reset_index().drop(columns= 'Day of Week').round(2)
df_revperday

,Day Name,Total_Bill
0,Monday,101677.28
1,Tuesday,99455.94
2,Wednesday,100313.54
3,Thursday,100767.78
4,Friday,101373.00
5,Saturday,96894.48
6,Sunday,98330.31


In [217]:
px.bar(data_frame=df_revperday, x='Day Name', y='Total_Bill', title='Total revenue per weekday', text_auto=True,
       labels={'Total_Bill': 'Total Revenue'}).show()

In [218]:
# To find the day with the highest revenue
df_revperday.sort_values(by= 'Total_Bill', ascending= False).head(1)

,Day Name,Total_Bill
0,Monday,101677.28


####  Which months generate the highest revenue?

In [219]:
df_revpermonth = df.groupby(['Month','Month Name'])['Total_Bill'].sum().reset_index().drop(columns= 'Month')
df_revpermonth

,Month Name,Total_Bill
0,January,81677.74
1,February,76145.19
2,March,98834.68
3,April,118941.08
4,May,156727.76
5,June,166485.88


In [220]:
px.bar(data_frame=df_revpermonth, x='Month Name', y='Total_Bill', title='Total revenue per month', text_auto=True, 
labels={'Total_Bill': 'Total Revenue'}).show()

- Conclusions: 
    - Mondays have the highest revenue per weekdays
    - Fridays have the highest revenue per weekends
    - June has the highest revenue

#### Which store has highest revenue?

In [221]:
df['store_location'].unique()

array(['Astoria', 'Lower Manhattan', "Hell's Kitchen"], dtype=object)

In [222]:
df_revperstore = df.groupby('store_location')['Total_Bill'].sum().reset_index().sort_values(by= 'Total_Bill', ascending= False)
df_revperstore

,store_location,Total_Bill
1,Hell's Kitchen,236511.17
0,Astoria,232243.91
2,Lower Manhattan,230057.25


In [223]:
px.pie(data_frame=df_revperstore, names= 'store_location', values= 'Total_Bill', title= 'Revenue per store location', hole=0.5).show()

- Hell's Kitchen has the highest revenue (all are approximately equal)

#### Which products sell most frequently?

In [224]:
df_popularProduct = df['product_category'].value_counts().reset_index() 
df_popularProduct

,product_category,count
0,Coffee,58416
1,Tea,45449
2,Bakery,22796
3,Drinking Chocolate,11468
4,Flavours,6790
5,Coffee beans,1753
6,Loose Tea,1210
7,Branded,747
8,Packaged Chocolate,487


In [225]:
# top 5 product category
px.histogram(data_frame=df_popularProduct.head(5), x='product_category', y='count', title= 'Top 5 product category', text_auto=True, labels={'product_category':'Products'}).show()

#### Most popular product category

In [226]:
px.pie(data_frame=df_popularProduct, names= 'product_category', values= 'count', title= 'Most popular product category', hole=0.5).show()

- Coffee is the most popular product

#### Product category and type distribution

In [227]:
df_chart = df.groupby(['product_category'])['product_type'].value_counts().reset_index()
df_chart

,product_category,product_type,count
0,Bakery,Scone,10173
1,Bakery,Pastry,6912
2,Bakery,Biscotti,5711
3,Branded,Housewares,526
4,Branded,Clothing,221
5,Coffee,Gourmet brewed coffee,16912
6,Coffee,Barista Espresso,16403
7,Coffee,Organic brewed coffee,8489
8,Coffee,Drip coffee,8477
9,Coffee,Premium brewed coffee,8135


In [228]:
px.bar(data_frame=df_chart, y='product_category', x='count', color= 'product_type', 
       title= 'Product category and type', text_auto=True,
       labels={'product_category':'Categories', 'count': 'Total number'}).show()

#### Revenue over time trends

In [229]:
df_daterev =  df.groupby(df['transaction_date'].dt.date)['Total_Bill'].sum().reset_index()
df_daterev

,transaction_date,Total_Bill
0,2023-01-01,2508.20
1,2023-01-02,2403.35
2,2023-01-03,2565.00
3,2023-01-04,2220.10
4,2023-01-05,2418.85
...,...,...
176,2023-06-26,5875.90
177,2023-06-27,5975.65
178,2023-06-28,4728.90
179,2023-06-29,4450.75


In [230]:
px.line(data_frame=df_daterev, x='transaction_date', y='Total_Bill', title= 'Revenue over time').show()

#### Trends of revenue per hour

In [231]:
df_salesperhour = df.groupby(df['Hour'])['Total_Bill'].sum().reset_index()
df_salesperhour

,Hour,Total_Bill
0,6,21900.27
1,7,63526.47
2,8,82699.87
3,9,85169.53
4,10,88673.39
5,11,46319.14
6,12,40192.79
7,13,40367.45
8,14,41304.74
9,15,41733.10


In [232]:
px.line(data_frame=df_salesperhour, x='Hour', y='Total_Bill', title='Trends of revenue per hour',labels={'Hour':'Clock Hour' , 'Total_Bill': 'Revenue per Hour'}).show()

#### Total Revenue by Day and Hour

In [233]:
sales_pivot = df.pivot_table(values='Total_Bill', index='Day Name', columns='Hour', aggfunc='sum')

#Order days correctly
day_order = [
   'Monday', 'Tuesday', 'Wednesday',
   'Thursday', 'Friday', 'Saturday', 'Sunday'
]

sales_pivot = sales_pivot.reindex(day_order)

# Plot heatmap
px.imshow(sales_pivot, text_auto=True, title='Total Revenue by Day and Hour').show()

- Range from 8:00 am till 10:00 am is the peak

# Feature Engineering

#### Average bill per time of day ('Morning', 'Afternoon', 'Evening', 'Night')

In [234]:
df['Day_time'] = df['Hour'].apply(lambda x: 'Morning' if  x < 12 else ('Afternoon' if  x < 16 else ('Evening' if  x < 21 else 'Night')))
df.head(3)

,transaction_id,transaction_date,store_location,transaction_qty,unit_price,Total_Bill,product_category,product_type,Size,Month Name,Day Name,Hour,Month,Day of Week,Day_time
0,114301,2023-06-01 11:33:29,Astoria,1,3.0,3.0,Tea,Brewed herbal tea,Large,June,Thursday,11,6,3,Morning
1,115405,2023-06-02 11:18:24,Astoria,1,3.0,3.0,Tea,Brewed herbal tea,Large,June,Friday,11,6,4,Morning
2,115478,2023-06-02 12:02:45,Astoria,1,3.0,3.0,Tea,Brewed herbal tea,Large,June,Friday,12,6,4,Afternoon


In [235]:
df_avgperdaytime = df.groupby('Day_time')['Total_Bill'].mean().round(2).reset_index()
df_avgperdaytime

,Day_time,Total_Bill
0,Afternoon,4.63
1,Evening,4.59
2,Morning,4.75


In [236]:
px.pie(data_frame=df_avgperdaytime, names= 'Day_time', values= 'Total_Bill', title= 'Average bill per time of day').show()

#### Product category per time of day

In [237]:
df_popcoffee = df.groupby(['Day_time', 'product_category'])['product_category'].count().reset_index(name='count').sort_values(['Day_time', 'count'], ascending=[True, False])
df_popcoffee

,Day_time,product_category,count
2,Afternoon,Coffee,14393
8,Afternoon,Tea,11225
0,Afternoon,Bakery,5000
4,Afternoon,Drinking Chocolate,2898
5,Afternoon,Flavours,1092
3,Afternoon,Coffee beans,282
6,Afternoon,Loose Tea,196
1,Afternoon,Branded,155
7,Afternoon,Packaged Chocolate,93
11,Evening,Coffee,13085


In [238]:
px.bar(data_frame=df_popcoffee, x='Day_time', y='count', color='product_category', barmode='group', title='Product category per time of day', labels={'Day_time': 'Time of Day', 'count': 'Number of Sales', 'product_category': 'Product Category'}).show()

#### Revenue for each day type [Weekends, Weekdays]

In [239]:
df['Day_Type'] = 'Weekday'
df.loc[df['Day of Week'].isin([4,5]), 'Day_Type'] = 'Weekend'
df.head()

,transaction_id,transaction_date,store_location,transaction_qty,unit_price,Total_Bill,product_category,product_type,Size,Month Name,Day Name,Hour,Month,Day of Week,Day_time,Day_Type
0,114301,2023-06-01 11:33:29,Astoria,1,3.0,3.0,Tea,Brewed herbal tea,Large,June,Thursday,11,6,3,Morning,Weekday
1,115405,2023-06-02 11:18:24,Astoria,1,3.0,3.0,Tea,Brewed herbal tea,Large,June,Friday,11,6,4,Morning,Weekend
2,115478,2023-06-02 12:02:45,Astoria,1,3.0,3.0,Tea,Brewed herbal tea,Large,June,Friday,12,6,4,Afternoon,Weekend
3,116288,2023-06-02 19:39:47,Astoria,1,3.0,3.0,Tea,Brewed herbal tea,Large,June,Friday,19,6,4,Evening,Weekend
4,116714,2023-06-03 12:24:57,Astoria,1,3.0,3.0,Tea,Brewed herbal tea,Large,June,Saturday,12,6,5,Afternoon,Weekend


In [240]:
df_revdaytype = df.groupby('Day_Type')['Total_Bill'].sum().reset_index()
df_revdaytype

,Day_Type,Total_Bill
0,Weekday,500544.85
1,Weekend,198267.48


In [241]:
px.bar(data_frame=df_revdaytype, x='Day_Type', y='Total_Bill', title='Revenue for each day type', text_auto= True ,labels={'Day_Type': 'Weekdays Vs Weekends', 'Total_Bill': 'Total Revenue'}).show()

# Conclusion

- The coffee shop generated a total revenue of $698,812 across 149,116 transactions, with June being the peak month ($166,486), Hell's Kitchen leading store performance ($236,511), and the 8–10 AM window consistently driving the highest daily sales across all locations and weekdays.


- Coffee dominated product demand (58,416 transactions), and since revenue was evenly spread across weekdays with Monday and Friday slightly ahead, the business should prioritize morning-hour promotions of coffee and top-selling bakery items to capitalize on proven peak-demand patterns and further grow profitability.

# Deployment

In [242]:
df.to_csv('Coffee_shop_sales_cleaned.csv', index= False)

In [ ]:
import streamlit as st

In [243]:
! streamlit run Home.py

^C
